File: python/glide-sync/glide_sync/client_pool.py
# Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

"""
Client-instance pool for the synchronous GLIDE client (Feature 1).

Backed by the shared Rust pool in glide-core via FFI. The Rust pool owns
client lifecycle (creation, LIFO reuse, bounded size). This Python
class provides the acquire-with-timeout retry loop and maps client_id handles
to usable GlideClient wrappers for command dispatch.

Usage:
    from glide_sync import GlideClient
    from glide_sync.client_pool import ClientPool, PoolConfig
    from glide_shared.config import GlideClientConfiguration, NodeAddress

    config = GlideClientConfiguration([NodeAddress("localhost", 6379)])
    pool = ClientPool(config, PoolConfig(max_size=10, min_idle=2))

    # Borrow a client, use it, return it automatically
    with pool.borrow() as client:
        client.set("key", "value")
        result = client.get("key")

    pool.close()
"""

import threading
import time
from dataclasses import dataclass
from typing import Optional

from glide_shared._glide_ffi import _GlideFFI
from glide_shared.config import BaseClientConfiguration, GlideClusterClientConfiguration

from .glide_client import BaseClient, GlideClient


@dataclass
class PoolConfig:
    """Configuration for the client-instance pool."""

    max_size: int = 10
    """Maximum number of clients in the pool."""

    min_idle: int = 1
    """Minimum idle clients to pre-warm at creation."""

    idle_timeout_ms: int = 300_000
    """Evict idle clients after this duration (ms). Default: 5 minutes."""

    request_timeout_ms: int = 5_000
    """Request timeout for commands (ms)."""

    acquire_timeout_s: float = 5.0
    """Maximum time to wait when pool is exhausted (seconds)."""


class ClientPool:
    """
    Client-instance pool backed by the Rust core.

    The Rust pool owns client lifecycle (creation, bounded size).
    This Python class provides:
    - acquire-with-timeout retry loop (exponential backoff)
    - Cached GlideClient wrappers (one per client_id, reused across borrows)
    - Zero-overhead borrow() context manager
    """

    __slots__ = (
        "_ffi",
        "_lib",
        "_client_config",
        "_pool_config",
        "_closed",
        "_client_cache",
        "_conn_req_bytes",
        "_pool_id",
        "_cache_lock",
    )

    def __init__(
        self,
        client_config: BaseClientConfiguration,
        pool_config: Optional[PoolConfig] = None,
    ):
        ffi_instance = _GlideFFI()
        self._ffi = ffi_instance.ffi
        self._lib = ffi_instance.lib
        self._client_config = client_config
        self._pool_config = pool_config or PoolConfig()
        self._closed = False
        self._client_cache: dict = {}
        # Lock for _client_cache under free-threading (concurrent get_or_create_client)
        self._cache_lock = threading.Lock()

        # Serialize the connection request protobuf
        is_cluster = isinstance(client_config, GlideClusterClientConfiguration)
        conn_req = client_config._create_a_protobuf_conn_request(
            cluster_mode=is_cluster
        )
        self._conn_req_bytes = conn_req.SerializeToString()

        # Create the Rust pool via FFI
        pool_id = self._lib.glide_pool_create(
            self._pool_config.max_size,
            self._pool_config.min_idle,
            self._pool_config.idle_timeout_ms,
            self._pool_config.request_timeout_ms,
            self._conn_req_bytes,
            len(self._conn_req_bytes),
        )

        if pool_id == -1:
            raise ValueError("Invalid pool configuration")
        if pool_id < 0:
            raise RuntimeError(f"Failed to create pool (error code: {pool_id})")

        self._pool_id = pool_id

    @property
    def pool_id(self) -> int:
        return self._pool_id

    @property
    def is_closed(self) -> bool:
        return self._closed

    def acquire(self, timeout: Optional[float] = None) -> int:
        """
        Acquire a client_id from the Rust pool.

        Retries with exponential backoff until a client is available or timeout.

        Returns:
            client_id (positive integer) for use with borrow() or get_client().

        Raises:
            TimeoutError: If no client available within timeout.
            RuntimeError: If pool is closed or invalid.
        """
        if self._closed:
            raise RuntimeError("Pool is closed")

        timeout = (
            timeout if timeout is not None else self._pool_config.acquire_timeout_s
        )
        deadline = time.monotonic() + timeout
        backoff_ms = 1.0

        while time.monotonic() < deadline:
            client_id = self._lib.glide_pool_try_acquire(self._pool_id)

            if client_id >= 0:
                return client_id
            if client_id == -2:
                raise RuntimeError("Invalid pool_id — pool was destroyed")

            remaining = deadline - time.monotonic()
            sleep_s = min(backoff_ms / 1000.0, max(remaining, 0))
            if sleep_s <= 0:
                break
            time.sleep(sleep_s)
            backoff_ms = min(backoff_ms * 2, 50.0)

        raise TimeoutError(
            f"Pool exhausted: could not acquire client within {timeout}s"
        )

    def release(self, client_id: int) -> None:
        """Release a borrowed client back to the pool."""
        self._lib.glide_pool_release(self._pool_id, client_id)

    def borrow(self, timeout: Optional[float] = None):
        """
        Returns a context manager that acquires a client from the Rust pool.

        The returned client dispatches commands through the native ClientAdapter
        managed by the Rust pool. On exit, the client is released back to the pool.

        Usage:
            with pool.borrow() as client:
                client.set("key", "value")
                result = client.get("key")
        """
        return _BorrowContext(self, timeout)

    def _get_or_create_client(self, client_id: int) -> BaseClient:
        """
        Get a usable GlideClient wrapper for the given client_id.

        On first borrow of a client_id, creates a thin wrapper pointing to the
        Rust-managed ClientAdapter. Subsequent borrows of the same client_id
        reuse the cached wrapper (zero allocation).
        """
        cached = self._client_cache.get(client_id)
        if cached is not None:
            return cached

        with self._cache_lock:
            # Double-check after acquiring lock
            cached = self._client_cache.get(client_id)
            if cached is not None:
                return cached

            # Get the native ClientAdapter pointer from the Rust pool
            adapter_ptr = self._lib.glide_pool_get_client_ptr(client_id)
            if adapter_ptr == 0:
                raise RuntimeError(
                    f"Pool client_id {client_id} has no associated ClientAdapter"
                )

            # Create a GlideClient shell pointing to the pooled adapter.
            cast_ptr = self._ffi.cast("void*", adapter_ptr)

            client = GlideClient.__new__(GlideClient)  # type: ignore[type-abstract]
            client._ffi = self._ffi
            client._lib = self._lib
            client._config = self._client_config
            client._is_closed = False
            client._pubsub_queue = []
            client._pubsub_lock = threading.Lock()
            client._pubsub_condition = threading.Condition(client._pubsub_lock)
            client._pubsub_callback_ref = None
            client._client_lock = threading.Lock()
            client._core_client = cast_ptr

            self._client_cache[client_id] = client
            return client

    def metrics(self) -> dict:
        """Get pool metrics: idle, active, total counts."""
        idle = self._ffi.new("uint32_t*")
        active = self._ffi.new("uint32_t*")
        total = self._ffi.new("uint32_t*")
        result = self._lib.glide_pool_metrics(self._pool_id, idle, active, total)
        if result != 0:
            return {"idle": 0, "active": 0, "total": 0}
        return {"idle": idle[0], "active": active[0], "total": total[0]}

    def close(self) -> None:
        """Destroy the pool. All idle clients are closed."""
        if not self._closed:
            self._closed = True
            self._lib.glide_pool_destroy(self._pool_id)
            self._client_cache.clear()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    def __del__(self):
        if not self._closed:
            try:
                self.close()
            except Exception:
                pass


class _BorrowContext:
    """Fast context manager for pool borrow/release (avoids generator overhead)."""

    __slots__ = ("_pool", "_timeout", "_client_id")

    def __init__(self, pool: ClientPool, timeout):
        self._pool = pool
        self._timeout = timeout
        self._client_id = -1

    def __enter__(self):
        self._client_id = self._pool.acquire(self._timeout)
        return self._pool._get_or_create_client(self._client_id)

    def __exit__(self, *_):
        self._pool.release(self._client_id)
        return False

File: java/client/src/main/java/glide/api/models/pool/ClientPool.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.pool;

import glide.api.GlideClient;
import glide.api.GlideClusterClient;
import glide.api.models.configuration.BaseClientConfiguration;
import glide.api.models.configuration.GlideClientConfiguration;
import glide.api.models.configuration.GlideClusterClientConfiguration;
import glide.api.models.configuration.ServerCredentials;
import glide.api.models.exceptions.ClosingException;
import glide.ffi.resolvers.GlidePoolResolver;
import glide.internal.GlideNativeBridge;
import glide.managers.ConnectionManager;
import java.time.Duration;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Client-instance pool backed by the shared Rust core.
 *
 * <p>Callers borrow a client via {@link #acquire()}, use it for commands, and return it via {@link
 * #release(long)}. The pool handles creation, LIFO reuse, bounded size, and background client
 * creation.
 *
 * <p>Usage:
 *
 * <pre>{@code
 * ClientPoolConfig config = ClientPoolConfig.builder()
 *     .maxSize(10)
 *     .minIdle(2)
 *     .clientConfig(GlideClientConfiguration.builder()
 *         .address(NodeAddress.builder().host("localhost").port(6379).build())
 *         .build())
 *     .build();
 *
 * ClientPool pool = ClientPool.create(config);
 * long clientId = pool.acquire().get();
 * GlideClient client = pool.getClient(clientId);
 * client.set("key", "value").get();
 * pool.release(clientId);
 * pool.close();
 * }</pre>
 */
public class ClientPool implements AutoCloseable {

    private static final int RUNNING = 0;
    private static final int CLOSED = 1;

    private final long poolId;
    private final ClientPoolConfig config;
    private final byte[] connectionRequestBytes;
    private final AtomicInteger state = new AtomicInteger(RUNNING);
    private final java.util.concurrent.ConcurrentHashMap<Long, GlideClient> clientCache =
            new java.util.concurrent.ConcurrentHashMap<>();

    private ClientPool(long poolId, ClientPoolConfig config, byte[] connectionRequestBytes) {
        this.poolId = poolId;
        this.config = config;
        this.connectionRequestBytes = connectionRequestBytes;
    }

    /**
     * Create a new pool.
     *
     * @param config pool configuration
     * @return the pool (clients are created asynchronously in the background)
     */
    public static ClientPool create(ClientPoolConfig config) {
        config.validate();

        // Reject pubsub subscriptions — pool state reset doesn't UNSUBSCRIBE
        if (config.getClientConfig().getSubscriptionConfiguration() != null) {
            throw new IllegalArgumentException(
                    "Pool clients cannot have pubsub subscriptions configured. "
                            + "Use the main client's pubsub API instead.");
        }

        // Reject a custom IAM credentials provider — it is a Java callback handed to native per
        // client, and glidePoolCreate takes only the request bytes with no way to forward it. Failing
        // here beats silently falling back to the core's default AWS credential chain (a different
        // principal).
        ServerCredentials credentials = config.getClientConfig().getCredentials();
        if (credentials != null
                && credentials.getIamConfig() != null
                && credentials.getIamConfig().getCredentialsProvider() != null) {
            throw new IllegalArgumentException(
                    "Pool clients cannot use a custom IAM credentials provider. "
                            + "Configure IAM without a provider to use the default credential chain.");
        }

        // Reject a custom address resolver for the same reason: it is a Java callback forwarded to
        // native per client, and glidePoolCreate cannot receive it. The connectivity probe below runs
        // the resolver and could pass, but pooled connections would then use the untranslated address
        // and fail at runtime — so fail fast here instead.
        if (config.getClientConfig().getAddressResolver().isPresent()) {
            throw new IllegalArgumentException(
                    "Pool clients cannot use a custom address resolver. "
                            + "Resolve addresses before configuring the pool.");
        }

        // Ahead of the connectivity probe below, so a static-config mistake surfaces as a
        // ConfigurationError naming the real reason instead of a probe failure. The serializer below
        // delegates to ConnectionManager's shared builder, so a pooled request carries the same fields
        // a directly-created client's does.
        ConnectionManager.validateClientAz(config.getClientConfig());

        byte[] connectionRequestBytes = serializeConnectionRequest(config.getClientConfig());

        long poolId =
                GlidePoolResolver.glidePoolCreate(
                        config.getMaxSize(),
                        config.getMinIdle(),
                        config.getIdleTimeout().toMillis(),
                        config.getRequestTimeout().toMillis(),
                        config.getAbandonTimeout().toMillis(),
                        connectionRequestBytes);

        if (poolId == -1) throw new IllegalArgumentException("Invalid pool configuration");
        if (poolId < 0) throw new RuntimeException("Pool creation failed: " + poolId);

        ClientPool pool = new ClientPool(poolId, config, connectionRequestBytes);

        // Connectivity probe: create one client to validate the config eagerly.
        // If this fails, propagate the actual connection error (not a timeout).
        try {
            AutoCloseable probeClient;
            if (config.getClientConfig() instanceof GlideClusterClientConfiguration) {
                probeClient =
                        GlideClusterClient.createClient(
                                        (GlideClusterClientConfiguration) config.getClientConfig())
                                .get(10, TimeUnit.SECONDS);
            } else {
                probeClient =
                        GlideClient.createClient((GlideClientConfiguration) config.getClientConfig())
                                .get(10, TimeUnit.SECONDS);
            }
            probeClient.close();
        } catch (Exception e) {
            pool.close();
            throw new RuntimeException("Pool connectivity probe failed", e);
        }

        return pool;
    }

    /** Get the native pool handle. */
    public long getPoolId() {
        return poolId;
    }

    /**
     * Acquire a pooled client with default timeout. Returns a pool-aware wrapper that returns the
     * client to the pool on close() (try-with-resources safe).
     */
    public CompletableFuture<PooledGlideClient> acquire() {
        return acquire(config.getAcquireTimeout());
    }

    /**
     * Acquire a pooled client with custom timeout. Retries with exponential backoff.
     *
     * <p>The returned {@link PooledGlideClient} implements AutoCloseable — its close() returns the
     * client to the pool instead of destroying the native connection.
     */
    public CompletableFuture<PooledGlideClient> acquire(Duration timeout) {
        if (state.get() != RUNNING) {
            CompletableFuture<PooledGlideClient> f = new CompletableFuture<>();
            f.completeExceptionally(new ClosingException("Pool is closed"));
            return f;
        }

        return CompletableFuture.supplyAsync(
                () -> {
                    long deadlineNanos = System.nanoTime() + timeout.toNanos();
                    long backoffMs = 1;

                    while (System.nanoTime() < deadlineNanos) {
                        if (state.get() != RUNNING) {
                            throw new RuntimeException(new ClosingException("Pool is closed"));
                        }

                        long clientId = GlidePoolResolver.glidePoolTryAcquire(poolId);
                        if (clientId >= 0) {
                            // test_on_borrow: PING to verify connection health
                            if (config.isTestOnBorrow()) {
                                try {
                                    GlideClient client = getClient(clientId);
                                    client.ping().get(2, java.util.concurrent.TimeUnit.SECONDS);
                                } catch (Exception e) {
                                    // Dead connection — release and retry
                                    release(clientId);
                                    continue;
                                }
                            }
                            return new PooledGlideClient(getClient(clientId), ClientPool.this, clientId);
                        }
                        if (clientId == -2) throw new RuntimeException("Pool was destroyed");

                        long remainingMs = (deadlineNanos - System.nanoTime()) / 1_000_000;
                        long sleepMs = Math.min(backoffMs, Math.max(remainingMs, 0));
                        if (sleepMs <= 0) break;
                        try {
                            Thread.sleep(sleepMs);
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                            throw new RuntimeException("Acquire interrupted", e);
                        }
                        backoffMs = Math.min(backoffMs * 2, 50);
                    }

                    throw new RuntimeException(
                            new java.util.concurrent.TimeoutException(
                                    "Pool exhausted: could not acquire within " + timeout));
                });
    }

    /**
     * Get a usable GlideClient for the given client_id. Cached — no allocation on subsequent calls
     * for the same client_id.
     *
     * <p><b>Important:</b> Do NOT call {@code close()} on the returned client. The pool manages the
     * client's lifecycle. Call {@link #release(long)} instead to return the client to the pool.
     */
    public GlideClient getClient(long clientId) {
        GlideClient cached = clientCache.get(clientId);
        if (cached != null) return cached;
        // Resolve the Java-side inflight limiter and request timeout the same way ConnectionManager
        // does for direct clients, so a pooled client fast-fails excess requests and times out
        // commands per its own config instead of the pool's defaults. The wire ConnectionRequest
        // already carries both values; these govern the host-side AsyncRegistry enforcement.
        //
        // AsyncRegistry keys its inflight counter on the raw native handle, and pooled client ids and
        // JNI handles come from independent counters that can collide, so a pooled client colliding
        // with a live ordinary client shares its counter and can be falsely rejected. The core limiter
        // enforces the real limit regardless, until the id spaces are unified.
        Integer configuredLimit = config.getClientConfig().getInflightRequestsLimit();
        int maxInflight =
                configuredLimit != null
                        ? configuredLimit
                        : GlideNativeBridge.getGlideCoreDefaultMaxInflightRequests();
        Integer configuredTimeout = config.getClientConfig().getRequestTimeout();
        long requestTimeoutMs =
                configuredTimeout != null
                        ? configuredTimeout
                        : GlideNativeBridge.getGlideCoreDefaultRequestTimeoutMs();
        // Carry the config's credentials so the borrowed client's IAM guards (refreshIamToken,
        // updateConnectionPassword) see them, as they do on a directly-created client.
        ServerCredentials credentials = config.getClientConfig().getCredentials();
        return clientCache.computeIfAbsent(
                clientId,
                id ->
                        GlideClient.fromPoolHandle(
                                id, maxInflight, requestTimeoutMs, credentials, connectionRequestBytes));
    }

    /** Release a client back to the pool. */
    public void release(long clientId) {
        GlidePoolResolver.glidePoolRelease(poolId, clientId);
    }

    /** Get pool metrics. */
    public int[] getMetrics() {
        return GlidePoolResolver.glidePoolMetrics(poolId);
    }

    public int getIdleCount() {
        int[] m = getMetrics();
        return m != null ? m[0] : 0;
    }

    public int getActiveCount() {
        int[] m = getMetrics();
        return m != null ? m[1] : 0;
    }

    public int getTotalCount() {
        int[] m = getMetrics();
        return m != null ? m[2] : 0;
    }

    @Override
    public void close() {
        if (state.compareAndSet(RUNNING, CLOSED)) {
            GlidePoolResolver.glidePoolDestroy(poolId);
            clientCache.clear();
        }
    }

    // ═══════════════════════════════════════════════════════════════════
    // Internal: protobuf serialization
    // ═══════════════════════════════════════════════════════════════════

    private static byte[] serializeConnectionRequest(BaseClientConfiguration config) {
        // Delegate to the one shared builder so a pooled client's request cannot drift from a
        // directly-created one.
        return ConnectionManager.buildConnectionRequest(config).toByteArray();
    }
}
